src/db.rs

```diff
@@ -2021,13 +2044,81 @@ mod tests {
         assert_eq!(snapshot.state.imm_memtable.get(1).unwrap().last_wal_id(), 2);
         assert_eq!(snapshot.state.core.next_wal_sst_id, next_wal_id);
         assert_eq!(
-            db.get(&key1).await.unwrap(),
+            reader.get(&key1).await.unwrap(),
             Some(Bytes::copy_from_slice(&value1))
         );
         assert_eq!(
-            db.get(&key2).await.unwrap(),
+            reader.get(&key2).await.unwrap(),
             Some(Bytes::copy_from_slice(&value2))
         );
+
+        fail_parallel::cfg(fp_registry.clone(), "write-compacted-sst-io-error", "off").unwrap();
+        db.close().await.unwrap();
+        reader.close().await.unwrap();
+    }
+
+    #[tokio::test]
+    async fn test_should_recover_imm_from_wal_after_flush_error() {
+        let fp_registry = Arc::new(FailPointRegistry::new());
+        fail_parallel::cfg(
+            fp_registry.clone(),
+            "write-compacted-sst-io-error",
+            "return",
+        )
+        .unwrap();
+
+        let object_store: Arc<dyn ObjectStore> = Arc::new(InMemory::new());
+        let path = Path::from("/tmp/test_kv_store");
+        let db = Db::open_with_fp_registry(
+            path.clone(),
+            test_db_options(0, 128, None),
+            object_store.clone(),
+            fp_registry.clone(),
+        )
+        .await
+        .unwrap();
+
+        // write a few keys that will result in memtable flushes
+        let key1 = [b'a'; 32];
+        let value1 = [b'b'; 96];
+        let result = db.put(&key1, &value1).await;
+        assert!(result.is_ok(), "Failed to write key1");
```

Comment:
This test has a race condition. It's not surfaced because the test runs in a single threaded runtime by default, so the memtable gets flushed before the second `put()` call happens. But if you set this:

`#[tokio::test(flavor = "multi_thread", worker_threads = 2)]`

Then the test fails. I believe this is because both `put()`'s execute successfully (they're durably written to the WAL) before the `mem_table_flush.rs` `select` loop receives and processes the `FlushImmutableMemtables` from `maybe_freeze_memtable`.

I am kinda thinking we should just remove this test since it seems to test almost the same thing as the test above--that we recover WAL files for kv pairs that haven't been written to L0 yet. But if you want to keep it, I think we need to figure out how to address the race condition--I'm nervous about depending on a single thread runtime to make the test work.

To make this test fail properly, I think you could drop the second put() all together, and instead, do:

```rust
        let key1 = [b'a'; 32];
        let value1 = [b'b'; 96];
        let result = db.put(&key1, &value1).await;
        assert!(result.is_ok(), "Failed to write key1");

        let flush_result = db.inner.flush_memtables().await;
        match flush_result {
            Err(e) => assert!(matches!(e, SlateDBError::IoError(_))),
            _ => panic!("Expected flush error"),
        }

        db.close().await.unwrap();
```

The `flush_memtables()` call should force the error to be triggered, thereby signaling that no SST was written to L0, as the test intends.

Reply:
yeah, that seems to work. I've updated this test with your suggestion. Thanks!